src/map/routing.c

```diff
@@ -168,24 +174,38 @@ static void ordered_enqueue(int next_offset, int current_dist, int remaining_dis
     ordered_queue_reduce_index(index, next_offset, possible_dist);
 }
 
-static inline int valid_offset(int grid_offset)
+static inline int valid_offset(int grid_offset, int possible_dist)
 {
-    return map_grid_is_valid_offset(grid_offset) && distance.determined.items[grid_offset] == 0;
+    int determined = distance.determined.items[grid_offset];
+    return map_grid_is_valid_offset(grid_offset) && (determined == 0 || possible_dist < determined);
 }
 
 static inline int distance_left(int x, int y)
 {
     return abs(distance.dst_x - x) + abs(distance.dst_y - y);
 }
 
+static int receive_highway_bonus(int offset, int direction)
+{
+    int terrain = map_terrain_get(offset);
+    if ((terrain & TERRAIN_HIGHWAY) == 0) {
```

Comment:
`if (!map_terrain_is(TERRAIN_HIGHWAY))`

Won't this work instead? Definitely unsure.

Reply:
I was really excited to use bitwise operators at the time.